### DISCUSSION TITLE
LaTeX is not rendered correctly in some cases

### DISCUSSION THREAD

**Discussion by AnotiaWang** [open-webui/open-webui#3581] [General]
# Bug Report

## Description

**Bug Summary:**
The UI cannot render LaTeX correctly when using LLM with Chinese. An expression can be shown only when it is enclosed with `[]`, like:
```
[ \mathbf{\hat{a}} = \frac{\mathbf{a}}{|\mathbf{a}|} ]
```

It won't work when `[` and `]` are on separate lines, also when expressions are enclosed with `$`.

**Steps to Reproduce:**
Ask a model to answer or explain a math problem in Chinese. You can also use the following output (edit a model's response and replace it):

<details>
<summary>Click to expand</summary>

```
平面、曲面和空间曲线是三维空间几何中的核心概念，它们的方程和性质是理解更复杂几何结构的基础。下面将逐一介绍这些概念及其相关知识点。

### 平面方程

平面的一般方程为：

[
Ax + By + Cz + D = 0
]

其中，$A$, $B$, $C$ 是平面的法向量的分量，$D$ 是常数项。

### 平面的夹角

两个平面的夹角由它们的法向量的夹角给出，设两个平面的法向量分别为 $\mathbf{n_1} = (A_1, B_1, C_1)$ 和 $\mathbf{n_2} = (A_2, B_2, C_2)$，则夹角 $\theta$ 由点积公式给出：

[ \cos \theta = \frac{\mathbf{n_1} \cdot \mathbf{n_2}}{\|\mathbf{n_1}\| \|\mathbf{n_2}\|} = \frac{A_1A_2 + B_1B_2 + C_1C_2}{\sqrt{A_1^2 + B_1^2 + C_1^2} \sqrt{A_2^2 + B_2^2 + C_2^2}} ]

### 曲面方程

曲面方程通常表示为 $F(x, y, z) = 0$ 的形式，其中 $F$ 是 $x$, $y$, $z$ 的函数。

### 二次曲面

二次曲面是 $F(x, y, z) = ax^2 + by^2 + cz^2 + dxy + exz + fyz + gx + hy + iz + j = 0$ 的形式，其中 $a$, $b$, $c$, $d$, $e$, $f$, $g$, $h$, $i$, $j$ 是常数。

常见的二次曲面包括椭球面、双曲面、抛物面等。

### 空间曲线

空间曲线可以由参数方程表示，如 $x = x(t)$, $y = y(t)$, $z = z(t)$，其中 $t$ 是参数。

### 解题思路：确定空间图形

1. **识别方程类型**：首先判断方程是平面方程、二次曲面方程还是空间曲线的参数方程。
2. **分析系数**：对于二次曲面方程，分析系数可以确定曲面的类型（如椭球面、双曲面等）。
3. **利用几何性质**：利用平面的法向量、曲面的对称性、曲线的参数范围等几何性质来进一步分析图形的特征。
4. **图形可视化**：使用数学软件或图形计算器可视化方程，帮助理解图形的形状和位置。

通过上述步骤，可以系统地分析和确定给定方程组描述的空间图形的性质和形状。
```

</details>

**Expected Behavior:**
The LaTeX in the response should be rendered. For example, here are two LaTeX expressions:
<img width="1112" alt="image" src="https://github.com/open-webui/open-webui/assets/62269186/2626f0b4-4511-449a-a31a-eae511572d78">

**Actual Behavior:**
For the example above, the renderer takes Chinese characters between two `$` as LaTeX.
<img width="1110" alt="image" src="https://github.com/open-webui/open-webui/assets/62269186/9e671351-d5fc-4e17-8ae6-9175db52bb45">

## Environment

- **Open WebUI Version:** 0.3.7
- **Ollama (if applicable):** None

- **Operating System:** macOS Sequoia
- **Browser (if applicable):** Vivaldi 6.8.3381.46 (Stable channel) (arm64) 

## Reproduction Details

**Confirmation:**

- [x] I have read and followed all the instructions provided in the README.md.
- [x] I am on the latest version of both Open WebUI and Ollama.
- [ ] I have included the browser console logs.
- [ ] I have included the Docker container logs.

## Logs and Screenshots

**Screenshots (if applicable):**
Rendered output for the example response:
![截屏2024-07-01 22 00 20](https://github.com/open-webui/open-webui/assets/62269186/664cab8a-76be-444e-a7bb-ce9dbcbd4a52)

## Installation Method

Docker

## Additional Information

None

## Note

If the bug report is incomplete or does not follow the provided instructions, it may not be addressed. Please ensure that you have followed the steps outlined in the README.md and troubleshooting.md documents, and provide all necessary information for us to reproduce and address the issue. Thank you!

**Comment by chrisoutwright** [comment]
I also get the issue with english:
it seems to depend on the model and input, but I see huge differences across the board:

![image](https://github.com/open-webui/open-webui/assets/27736055/6deda734-ec26-42bc-83d9-0a8d5868a061)
IS this now reponse prompt issue or how the library works for open-webui?

**Reply by AnotiaWang** [reply]
I guess open-webui's LaTeX engine only supports certain LaTeX format. I had to tell models to write expressions and symbols using `[]`.

**Reply by chrisoutwright** [reply]
would be good to have a consistent expectation per model  in how this is output (usual markdown) or some strategies in openwebui to cope with syntax (proposal to have a small model/algo rectifying issue with a button after response generation). How does GPT4 manage it? Has it been heavily trained as for not alternating notation? 
At the moment [ ] across lines in the biggest issue together with missing inline often. Often the model will answer in style of input and use some mix.

**Comment by chrisoutwright** [comment]
I tried the latest openwebui and it seems still that inline is persistently not formatted, here the output on copy:
```
You're right to ask! If there's no bias, meaning both dice behave identically (both are fair), then the KL divergence will be zero.

Let's see why:

* **Identical Distributions**: When \( p = q \) (meaning the probability distribution of rolling each face is the same for both dice), then \( \frac{p(x)}{q(x)} = 1 \).
* **Logarithm of 1**:  The logarithm of 1 (log 1) is always zero.

Therefore, in the formula:

\[ D_{KL}(p||q) = \sum_x p(x)\log\frac{p(x)}{q(x)} \]

Each term \( p(x)\log\frac{p(x)}{q(x)} \) will be zero since \( \frac{p(x)}{q(x)} = 1 \). The sum of zeros is still zero.


**Conclusion:** When the two distributions are identical (like two fair dice), the KL divergence between them is zero, indicating no information gain from switching between them.
```
Image:
![SmartSelect_20240715_004803_Chrome](https://github.com/user-attachments/assets/e5a510ee-0a23-4a18-8ffc-fe2ad3f4e684)
 

There is even for the parentheses escaped  the extra whitespace but still..

**Reply by chrisoutwright** [reply]
Is this not rather easy to fix?

https://github.com/open-webui/open-webui/blob/eff736acd2e0bbbdd0eeca4cc209b216a1f23b6a/src/lib/components/chat/Messages/ResponseMessage.svelte#L173

Is it missing in notation \\( ... \\)  ?

**Reply by justinh-rahb** [reply]
> Is this not rather easy to fix?

You'd think so, but this has been a persistent issue since we introduced LaTeX rendering to begin with, and we've flipped back and forth between liberally applying it to anything that even looks like LaTeX, to trying to be more selective, all at the behest of issues reports... so it seems to me now that there is no **ONE** way to do this satisfactorily.

**Reply by chrisoutwright** [reply]
@justinh-rahb can you point out to me those examples where it was identifying it wrongly? Technically what has been the culprit there? A "\\( " and a closing "\\)" seem like not being used that often in natural speech , more like in a regex itself being stated, but these would be within a code formatting and could be the excluded. I would like to understand where we are stuck and which solution could really be done. At the moment the chat experience is being detracted for me a bit as mathematically inclined person.

**Reply by Yanyutin753** [reply]
https://github.com/open-webui/open-webui/pull/3930 Look at this. I tried to solve this problem

**Reply by Yanyutin753** [reply]
@chrisoutwright  https://github.com/open-webui/open-webui/pull/3930 Look at this. I tried to solve this problem.

**Comment by jiangying000** [comment]
This also often occurs in my case, such as:


[
\langle f, g \rangle = \int_{-\pi}^{\pi} f(x) g(x) , dx
]
